sql: TestRelocateNonVoters failed [trying to add({ChangeType:ADD_NON_VOTER...}) to a store that already has a LEARNER]

sql.TestRelocateNonVoters [failed](https://tanzanite.cluster.engflow.com/invocations/default/17c023ff-a71d-4307-a445-7ac16cedbefa?testReportRun=18&testReportShard=15&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [7c93a5c7f2e3e12b4923fbca4e26c284eb776aa3](https://github.com/cockroachdb/cockroach/commits/7c93a5c7f2e3e12b4923fbca4e26c284eb776aa3):


```
=== RUN   TestRelocateNonVoters
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRelocateNonVoters119892935
    test_log_scope.go:81: use -show-logs to present logs inline
    multitenant_admin_function_test.go:956: -- test log scope end --
test logs left over in: outputs.zip/logTestRelocateNonVoters119892935
--- FAIL: TestRelocateNonVoters (67.59s)
=== RUN   TestRelocateNonVoters/ALTER_RANGE_x_RELOCATE_NONVOTERS
    multitenant_admin_function_test.go:377: condition failed to evaluate within 45s: from multitenant_admin_function_test.go:414: expected "trying to add({ChangeType:ADD_NON_VOTER Target:n1,s1}) to a store that already has a LEARNER" contains "ok" tenant=secondary query=``ALTER RANGE (SELECT min(range_id) FROM [SHOW RANGES FROM TABLE t]) RELOCATE NONVOTERS FROM 4 TO 1;`` leaseholder=5 replicas=[2 3 4 5] voting_replicas=[2 3 5] non_voting_replicas=[4] fromReplica=4 toReplica=1 row=0 col=2
    --- FAIL: TestRelocateNonVoters/ALTER_RANGE_x_RELOCATE_NONVOTERS (58.81s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=18</code>
 - <code>shard=15</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRelocateNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39968

sql.TestRelocateNonVoters [failed](https://tanzanite.cluster.engflow.com/invocations/default/09d589ce-cd76-429d-99af-be5c713ed93c?testReportRun=9&testReportShard=15&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [83589fb87caa92fb42e83994f1691978f37e4cbb](https://github.com/cockroachdb/cockroach/commits/83589fb87caa92fb42e83994f1691978f37e4cbb):


```
=== RUN   TestRelocateNonVoters
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRelocateNonVoters440182711
    test_log_scope.go:81: use -show-logs to present logs inline
    multitenant_admin_function_test.go:956: -- test log scope end --
test logs left over in: outputs.zip/logTestRelocateNonVoters440182711
--- FAIL: TestRelocateNonVoters (70.32s)
=== RUN   TestRelocateNonVoters/ALTER_RANGE_RELOCATE_NONVOTERS
    multitenant_admin_function_test.go:377: condition failed to evaluate within 45s: from multitenant_admin_function_test.go:414: expected "trying to add({ChangeType:ADD_NON_VOTER Target:n1,s1}) to a store that already has a LEARNER" contains "ok" tenant=secondary query=``ALTER RANGE RELOCATE NONVOTERS FROM 4 TO 1 FOR (SELECT min(range_id) FROM [SHOW RANGES FROM TABLE t]);`` leaseholder=3 replicas=[2 3 4 5] voting_replicas=[2 3 5] non_voting_replicas=[4] fromReplica=4 toReplica=1 row=0 col=2
    --- FAIL: TestRelocateNonVoters/ALTER_RANGE_RELOCATE_NONVOTERS (59.18s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=9</code>
 - <code>shard=15</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRelocateNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

sql.TestRelocateNonVoters [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17043297?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17043297?buildTab=artifacts#/) on master @ [08e000634f5ccade168fc8e160eb0310d2b25426](https://github.com/cockroachdb/cockroach/commits/08e000634f5ccade168fc8e160eb0310d2b25426):


```
=== RUN   TestRelocateNonVoters
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestRelocateNonVoters362562139
    test_log_scope.go:81: use -show-logs to present logs inline
    multitenant_admin_function_test.go:956: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestRelocateNonVoters362562139
--- FAIL: TestRelocateNonVoters (71.05s)
=== RUN   TestRelocateNonVoters/ALTER_RANGE_RELOCATE_NONVOTERS
    multitenant_admin_function_test.go:377: condition failed to evaluate within 45s: from multitenant_admin_function_test.go:414: expected "trying to add({ChangeType:ADD_NON_VOTER Target:n2,s2}) to a store that already has a LEARNER" contains "ok" tenant=secondary query=``ALTER RANGE RELOCATE NONVOTERS FROM 5 TO 2 FOR (SELECT min(range_id) FROM [SHOW RANGES FROM TABLE t]);`` leaseholder=4 replicas=[1 3 4 5] voting_replicas=[1 3 4] non_voting_replicas=[5] fromReplica=5 toReplica=2 row=0 col=2
    --- FAIL: TestRelocateNonVoters/ALTER_RANGE_RELOCATE_NONVOTERS (58.02s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRelocateNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

sql.TestRelocateNonVoters [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17127016?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17127016?buildTab=artifacts#/) on master @ [ec2573dc6aaeefc226440bb2c5a7c94a63989868](https://github.com/cockroachdb/cockroach/commits/ec2573dc6aaeefc226440bb2c5a7c94a63989868):


```
=== RUN   TestRelocateNonVoters
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestRelocateNonVoters4102263187
    test_log_scope.go:81: use -show-logs to present logs inline
    multitenant_admin_function_test.go:956: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestRelocateNonVoters4102263187
--- FAIL: TestRelocateNonVoters (71.55s)
=== RUN   TestRelocateNonVoters/ALTER_RANGE_RELOCATE_NONVOTERS
    multitenant_admin_function_test.go:377: condition failed to evaluate within 45s: from multitenant_admin_function_test.go:414: expected "trying to add({ChangeType:ADD_NON_VOTER Target:n5,s5}) to a store(5) that already has a replica(VOTER_FULL)" contains "ok" tenant=system query=``ALTER RANGE RELOCATE NONVOTERS FROM 4 TO 5 FOR (SELECT min(range_id) FROM [SHOW RANGES FROM TABLE t]);`` leaseholder=5 replicas=[1 2 3 4] voting_replicas=[1 2 3] non_voting_replicas=[4] fromReplica=4 toReplica=5 row=0 col=2
    --- FAIL: TestRelocateNonVoters/ALTER_RANGE_RELOCATE_NONVOTERS (57.81s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #131653 sql: TestRelocateNonVoters failed [C-test-failure O-robot T-sql-queries branch-release-24.2.3-rc release-blocker]
- #131436 sql: TestRelocateNonVoters failed [C-test-failure O-robot T-sql-foundations branch-release-24.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRelocateNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/sql/multitenant_admin_function_test.go
+++ b/pkg/sql/multitenant_admin_function_test.go
@@ -17,6 +17,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/config/zonepb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
 	"github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/server"
@@ -56,6 +57,7 @@ func createTestClusterArgs(ctx context.Context, numReplicas, numVoters int32) ba
 
 	clusterSettings := cluster.MakeTestingClusterSettings()
 	kvserver.LoadBasedRebalancingMode.Override(ctx, &clusterSettings.SV, kvserver.LBRebalancingOff)
+	kvserverbase.MergeQueueEnabled.Override(ctx, &clusterSettings.SV, false)
 	return base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
 			Settings: clusterSettings,
@@ -747,6 +749,7 @@ func TestRelocateVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
+					testCluster.ToggleLeaseQueues(false)
 					testCluster.ToggleReplicateQueues(false)
 					testCluster.ToggleSplitQueues(false)
 					replicaState := getReplicaState(
@@ -827,6 +830,7 @@ func TestExperimentalRelocateVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
+					testCluster.ToggleLeaseQueues(false)
 					testCluster.ToggleReplicateQueues(false)
 					testCluster.ToggleSplitQueues(false)
 					replicaState := getReplicaState(
@@ -920,6 +924,7 @@ func TestRelocateNonVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
+					testCluster.ToggleLeaseQueues(false)
 					testCluster.ToggleReplicateQueues(false)
 					testCluster.ToggleSplitQueues(false)
 					replicaState := getReplicaState(
@@ -994,6 +999,7 @@ func TestExperimentalRelocateNonVoters(t *testing.T) {
 					require.NoErrorf(t, err, message)
 					err = testCluster.WaitForFullReplication()
 					require.NoErrorf(t, err, message)
+					testCluster.ToggleLeaseQueues(false)
 					testCluster.ToggleReplicateQueues(false)
 					testCluster.ToggleSplitQueues(false)
 					replicaState := getReplicaState(

--- a/pkg/testutils/serverutils/test_cluster_shim.go
+++ b/pkg/testutils/serverutils/test_cluster_shim.go
@@ -267,6 +267,10 @@ type TestClusterInterface interface {
 	// TogglesplitQueues activates or deactivates the split queues on all
 	// the stores on all the nodes.
 	ToggleSplitQueues(active bool)
+
+	// ToggleLeaseQueues activates or deactivates the lease queues on all
+	// the stores on all the nodes.
+	ToggleLeaseQueues(active bool)
 }
 
 // SplitPoint describes a split point that is passed to SplitTable.

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -1606,6 +1606,16 @@ func (tc *TestCluster) ToggleSplitQueues(active bool) {
 	}
 }
 
+// ToggleLeaseQueues implements TestClusterInterface.
+func (tc *TestCluster) ToggleLeaseQueues(active bool) {
+	for _, s := range tc.Servers {
+		_ = s.StorageLayer().GetStores().(*kvserver.Stores).VisitStores(func(store *kvserver.Store) error {
+			store.TestingSetLeaseQueueActive(active)
+			return nil
+		})
+	}
+}
+
 // ReadIntFromStores reads the current integer value at the given key
 // from all configured engines on un-stopped servers, filling in zeros
 // when the value is not found.

